File: backend/app/schemas/course.py
from typing import Optional
from pydantic import BaseModel, ConfigDict


class CourseSummaryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    source_language: str
    target_language: str
    description: Optional[str] = None
    order_index: int
    is_active: bool


class LessonSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    skill_id: int
    title: str
    order_index: int
    xp_reward: int
    is_completed: bool = False


class SkillMapNode(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    unit_id: int
    title: str
    subtitle: Optional[str] = None
    description: Optional[str] = None
    icon_key: str
    order_index: int
    xp_reward: int
    is_locked_by_default: bool
    status: str  # "locked" | "unlocked" | "in_progress" | "completed" | "mastered"
    is_unlocked: bool
    completed: bool
    crown_level: int
    lessons_completed: int
    total_lessons: int
    xp_earned: int
    lessons: list[LessonSummary]


class UnitMapSection(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    course_id: int
    title: str
    description: Optional[str] = None
    order_index: int
    skills: list[SkillMapNode]


class LoopMapResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    course_id: int
    course_name: str
    source_language: str
    target_language: str
    description: Optional[str] = None
    units: list[UnitMapSection]


class SkillDetailResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    unit_id: int
    unit_title: str
    title: str
    subtitle: Optional[str] = None
    description: Optional[str] = None
    icon_key: str
    order_index: int
    xp_reward: int
    status: str
    is_unlocked: bool
    completed: bool
    crown_level: int
    lessons_completed: int
    total_lessons: int
    xp_earned: int
    lessons: list[LessonSummary]

File: frontend/lib/api/course.ts
import { apiFetch } from "./client";

export interface CourseSummary {
  id: number;
  name: string;
  source_language: string;
  target_language: string;
  description: string | null;
  order_index: number;
  is_active: boolean;
}

export interface LessonSummary {
  id: number;
  skill_id: number;
  title: string;
  order_index: number;
  xp_reward: number;
  is_completed: boolean;
}

export interface SkillMapNode {
  id: number;
  unit_id: number;
  title: string;
  subtitle: string | null;
  description: string | null;
  icon_key: string;
  order_index: number;
  xp_reward: number;
  is_locked_by_default: boolean;
  status: "locked" | "unlocked" | "in_progress" | "completed" | "mastered" | string;
  is_unlocked: boolean;
  completed: boolean;
  crown_level: number;
  lessons_completed: number;
  total_lessons: number;
  xp_earned: number;
  lessons: LessonSummary[];
}

export interface UnitMapSection {
  id: number;
  course_id: number;
  title: string;
  description: string | null;
  order_index: number;
  skills: SkillMapNode[];
}

export interface LoopMap {
  course_id: number;
  course_name: string;
  source_language: string;
  target_language: string;
  description: string | null;
  units: UnitMapSection[];
}

export interface SkillDetail {
  id: number;
  unit_id: number;
  unit_title: string;
  title: string;
  subtitle: string | null;
  description: string | null;
  icon_key: string;
  order_index: number;
  xp_reward: number;
  status: string;
  is_unlocked: boolean;
  completed: boolean;
  crown_level: number;
  lessons_completed: number;
  total_lessons: number;
  xp_earned: number;
  lessons: LessonSummary[];
}

/**
 * Fetches the active language course metadata.
 * Target: GET /api/courses/active
 */
export async function getActiveCourse(): Promise<CourseSummary> {
  return apiFetch<CourseSummary>("/api/courses/active", {
    method: "GET",
    cache: "no-store",
  });
}

/**
 * Fetches the complete Loop Map hierarchy with learner progress.
 * Target: GET /api/courses/{courseId}/map
 */
export async function getLoopMap(courseId: number): Promise<LoopMap> {
  return apiFetch<LoopMap>(`/api/courses/${courseId}/map`, {
    method: "GET",
    cache: "no-store",
  });
}

/**
 * Fetches full details for a specific skill.
 * Target: GET /api/skills/{skillId}
 */
export async function getSkillDetail(skillId: number): Promise<SkillDetail> {
  return apiFetch<SkillDetail>(`/api/skills/${skillId}`, {
    method: "GET",
    cache: "no-store",
  });
}
